### PR TITLE
DR-1441 Upgrade openapi (with new gradle version)

### PR DESCRIPTION
**Pull request by myessail**

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -24,6 +24,6 @@ indent_size=2
 indent_style=space
 indent_size=2
 
-[{*.yml,*.yaml}]
+[{*.yml,*.yaml,*.gradle}]
```
Comment (pshapiro4broad) [inline .editorconfig:27]: Is there a reason to have three different groups that all have the same indent style settings? Also why do they all specify `indent_style=space` when that's the setting used in the `*` config?

File: .editorconfig [hunk 2]
```diff
@@ -24,6 +24,6 @@ indent_size=2
...
 indent_style=space
-indent_size=2
+indent_size=2
```
Comment (nmalfroy) [inline .editorconfig:29]: This should have been pulled into a separate PR I think or have meaningful commits.  Please try to avoid mixing whitespace changes with code changes

File: build.gradle [hunk 3]
```diff
@@ -11,344 +11,333 @@
...
+  id "checkstyle"
 }
 
 allprojects {
-    group 'bio.terra'
-    version '1.0.146-SNAPSHOT'
-}
-
-// skip subproject tasks by default when building jade-data-repo
-subprojects.each { s ->
```
Comment (nmalfroy) [inline build.gradle]: Question for the team: was there any reason for this to be here?

File: src/main/java/bio/terra/app/controller/UnauthenticatedApiController.java [hunk 4]
```diff
@@ -115,14 +114,4 @@ public UnauthenticatedApiController(
         }
         return new ResponseEntity<>(HttpStatus.NO_CONTENT);
     }
-
-    /**
-     * Home redirection to swagger api documentation
-     */
-    @RequestMapping(value = "/")
```
Comment (ddietterich) [inline src/main/java/bio/terra/app/controller/UnauthenticatedApiController.java]: How do we resolve to the swagger page without this?

File: build.gradle [hunk 5]
```diff
@@ -11,344 +11,333 @@
...
-    repositories {
-        mavenCentral()
-    }
-    dependencies {
-        classpath("org.springframework.boot:spring-boot-gradle-plugin:${springBootVersion}")
-        classpath('io.swagger:swagger-codegen:2.4.11')
-        classpath('gradle.plugin.com.srcclr:gradle:3.1.0')
-    }
+  ext {
+    springBootVersion = '2.3.4.RELEASE'
```
Comment (ddietterich) [inline build.gradle]: What motivated the indentation change? Ideally, it would be good to do that as a separate step, since now the whole file is different.

File: build.gradle [hunk 6]
```diff
@@ -11,344 +11,333 @@
...
 
-apply plugin: 'java'
-apply plugin: 'idea'
-apply plugin: 'org.springframework.boot'
-apply plugin: 'io.spring.dependency-management'
-apply plugin: 'antlr'
-
-apply plugin: 'checkstyle'
-apply plugin: 'jacoco'
 apply plugin: 'com.github.spotbugs'
```
Comment (ddietterich) [inline build.gradle]: Can't these plugins also use the new plugin style?

File: build.gradle [hunk 7]
```diff
@@ -11,344 +11,333 @@
...
+  }
 }
 
 configurations {
-    generatedCompile
-    runtimeClasspath
-}
-
-// Exclude the Spring logger, so everything will use SLF4J
-configurations.all {
```
Comment (nmalfroy) [inline build.gradle]: Why are we getting rid of this?

Comment (ddietterich) [inline build.gradle]: This was required to get the right logging out of Spring. Why can it be removed?

File: src/main/resources/api/swagger-ui.html [hunk 8]
```diff
@@ -0,0 +1,98 @@
+<!-- Copied from the swagger-ui static file -->
```
Comment (ddietterich) [inline src/main/resources/api/swagger-ui.html:1]: What happens if a new version of the code generator shows up and changes this file?
Or asked another way, why is it safe to drop this file into the source?

**Review by ddietterich — COMMENTED** [review]
I don't have any changes to request, but I do have lots of questions.

File: build.gradle (hunk 5 above)
Comment (pshapiro4broad) [reply build.gradle]: I agree, although now that it's done, using "hide whitespace changes" in github can help.

File: config/checkstyle/checkstyle-suppressions.xml [hunk 9]
```diff
@@ -5,4 +5,5 @@
 
 <suppressions>
     <suppress checks="[a-zA-Z0-9]*" files="[\\/]grammar[\\/]" />
+    <suppress files="[\\/]generated-src[\\/]" checks="[a-zA-Z0-9]*"/>
```
Comment (RoriCremer) [inline config/checkstyle/checkstyle-suppressions.xml]: were we checking generated code before? I assume we were not, but maybe this is in a new location? It would seem odd to me to have more than one place for checkstyle suppression

**Review by RoriCremer — APPROVED** [review]

File: src/main/resources/api/data-repository-openapi.yaml [hunk 10]
```diff
@@ -0,0 +1,3638 @@
...
+     3. A separate part of the definitions section for the DRS definitions
+
+    NOTE: the code here does not relect the DRS spec anymore. See DR-409.
+  license:
+    name: Apache 2.0
+    url: https://www.apache.org/licenses/LICENSE-2.0.html
+  version: 0.1.0
+servers:
+  - url: /
+security:
```
Comment (pshapiro4broad) [inline src/main/resources/api/data-repository-openapi.yaml]: Some comments appear to have been removed, was that intentional? In the original yaml, there was a comment before this line:
```
##########################################################################################
# AUTH
##########################################################################################
```

File: src/main/resources/api/data-repository-openapi.yaml [hunk 11]
```diff
@@ -0,0 +1,3638 @@
...
+  - googleoauth:
+      - openid
+      - email
+      - profile
+paths:
+  /status:
```
Comment (pshapiro4broad) [inline src/main/resources/api/data-repository-openapi.yaml]: and before here:
```
  ## Unauthenticated endpoints
```

File: datarepo-client/build.gradle [hunk 12]
```diff
@@ -40,64 +25,43 @@ repositories {
...
+                "--invoker-package", "bio.terra.datarepo.client",
+                "--group-id", "bio.terra.datarepo",
+                "--artifact-version", "${version}",
+                "--ignore-file-override", "${projectDir}/.swagger-codegen-ignore",
+                "-D", "apiTests=false," +
+                    "apiDocs=false," +
+                    "modelTests=false," +
+                    "modelDocs=false," +
+                    "dateLibrary=java8"
+            ]
```
Comment (RoriCremer) [inline datarepo-client/build.gradle]: do we lose params like:
interfaceOnly
invokerPackage
etc?

File: datarepo-client/build.gradle [hunk 13]
```diff
@@ -125,20 +89,3 @@ artifactory {
...
-    delete "README.md"
-    delete "build.sbt"
-    delete "docs/"
-    delete "git_push.sh"
-    delete "gradle.properties"
-    delete "gradle/"
-    delete "gradlew"
-    delete "gradlew.bat"
-    delete "pom.xml"
-    delete "src/"
```
Comment (RoriCremer) [inline datarepo-client/build.gradle]: are these still getting removed?

File: src/main/java/bio/terra/service/dataset/DatasetRequestValidator.java [hunk 14]
```diff
@@ -90,14 +90,6 @@ boolean isValidRelationship(String relationshipName) {
         }
     }
 
-    private void validateDatasetName(String datasetName, Errors errors) {
-        // NOTE: We used to manually check the name against a pattern here, but the latest
-        // versions of Swagger codegen now auto-generate an equivalent check.
-        if (datasetName == null) {
-            errors.rejectValue("name", "DatasetNameMissing");
-        }
-    }
```
Comment (RoriCremer) [inline src/main/java/bio/terra/service/dataset/DatasetRequestValidator.java:99]: nit: wouldn't it be more descriptive to have this error over "NotNull" ?

File: src/main/java/bio/terra/app/configuration/SwaggerDocumentationConfig.java [hunk 15]
```diff
@@ -1,145 +0,0 @@
...
-
-    @Autowired
-    public SwaggerDocumentationConfig(OauthConfiguration oauthConfig) {
-        this.oauthConfig = oauthConfig;
-    }
-
-    ApiInfo apiInfo() {
-        return new ApiInfoBuilder()
-            .title("Data Repository API")
-            .description("This document defines the REST API for Data Repository.\n\n" +
```
Comment (nmalfroy) [inline src/main/java/bio/terra/app/configuration/SwaggerDocumentationConfig.java:47]: Where does all of this come from then?  Are we losing our swagger page documentation?

File: src/main/java/bio/terra/app/controller/UnauthenticatedApiController.java [hunk 16]
```diff
@@ -115,14 +114,4 @@ public UnauthenticatedApiController(
...
-    public String index() {
```
Comment (nmalfroy) [inline src/main/java/bio/terra/app/controller/UnauthenticatedApiController.java]: What happens when we we the root endpoint now?

File: src/main/resources/api/data-repository-openapi.yaml [hunk 17]
```diff
@@ -0,0 +1,3638 @@
+openapi: 3.0.1
```
Comment (pshapiro4broad) [inline src/main/resources/api/data-repository-openapi.yaml]: It looks like the current version is 3.0.3, would it be worth changing to that version now?

File: src/main/java/bio/terra/service/dataset/DatasetRequestValidator.java [hunk 18]
```diff
@@ -90,14 +90,6 @@ boolean isValidRelationship(String relationshipName) {
         }
     }
 
-    private void validateDatasetName(String datasetName, Errors errors) {
-        // NOTE: We used to manually check the name against a pattern here, but the latest
```
Comment (nmalfroy) [inline src/main/java/bio/terra/service/dataset/DatasetRequestValidator.java:94]: Is this because the new code gen handles this for us?

File: src/main/resources/api/data-repository-openapi.yaml (hunk 17 above)
Comment (nmalfroy) [inline src/main/resources/api/data-repository-openapi.yaml]: Any sense of how much this differs from our current yaml?

File: src/main/resources/api/data-repository-openapi.yaml [hunk 19]
```diff
@@ -0,0 +1,3638 @@
...
+                $ref: '#/components/schemas/ErrorModel'
+      x-codegen-request-body-name: billingProfileRequest
+  /api/resources/v1/profiles/{id}:
+    get:
+      tags:
+        - resources
+      description: Retrieve a billing profile by id
+      operationId: retrieveProfile
+      parameters:
+        - name: id
```
Comment (pshapiro4broad) [inline src/main/resources/api/data-repository-openapi.yaml]: These and a number of other `id`s used to be `- $ref: '#/parameters/Id'`. Is there no way to use `$ref` for these id fields anymore?

File: src/main/resources/application.properties [hunk 20]
```diff
@@ -5,7 +5,9 @@ spring.datasource.tomcat.validation-query=SELECT 1
 spring.jackson.serialization.WRITE_DATES_AS_TIMESTAMPS=false
 spring.liquibase.enabled=false
 spring.profiles.active=google
-springfox.documentation.swagger.v2.path=/api-docs
+spring.resources.cache.cachecontrol.max-age=0
+spring.resources.cache.cachecontrol.must-revalidate=true
+spring.resources.static-locations=classpath:/api/
```
Comment (nmalfroy) [inline src/main/resources/application.properties]: This probably isn't a great path of static resources

File: src/main/resources/api/data-repository-openapi.yaml [hunk 21]
```diff
@@ -0,0 +1,3638 @@
...
+              schema:
+                $ref: '#/components/schemas/ErrorModel'
+  /api/repository/v1/snapshots/{id}/filesystem/objects:
+    get:
+      tags:
+        - repository
+      description: Lookup metadata for one file
+      operationId: lookupSnapshotFileByPath
+      parameters:
+        - name: id
```
Comment (pshapiro4broad) [inline src/main/resources/api/data-repository-openapi.yaml]: In the old yaml this was:
```
      parameters:
        - $ref: '#/parameters/Id'
        - $ref: '#/parameters/Path'
        - $ref: '#/parameters/Depth'
```

**Review by nmalfroy — COMMENTED** [review]
Same as dan, no specific change requests but some questions.  This likely warrants `skaffold run`-ing this to verify

File: src/main/resources/api/data-repository-openapi.yaml [hunk 22]
```diff
@@ -0,0 +1,3638 @@
+openapi: 3.0.1
+info:
```
Comment (pshapiro4broad) [inline src/main/resources/api/data-repository-openapi.yaml]: This is a huge file to look at! Maybe in the future, and if there's a way to do it, the yaml can be split into different files based on endpoints (using an include mechanism?) to make it easier to work on.

File: build.gradle (hunk 3 above)
Comment (fboulnois) [reply build.gradle]: Yes. @ddietterich had a concern about building subprojects every time the main project is built. I don't think it is desirable to remove this unless there is team agreement that it should be done.

File: datarepo-client/build.gradle (hunk 13 above)
Comment (fboulnois) [reply datarepo-client/build.gradle]: I think this task should be retained.

File: datarepo-client/build.gradle [hunk 23]
```diff
@@ -40,64 +25,43 @@ repositories {
...
-
-    testCompile group: 'junit', name: 'junit', version: '4.12'
-    testCompile group: 'org.hamcrest', name: 'hamcrest', version: '2.1'
+    // Swagger deps, versions controlled by dependencyManagement in root project
+    implementation group: "org.glassfish.jersey.core", name: "jersey-client"
+    implementation group: "org.glassfish.jersey.media", name: "jersey-media-json-jackson"
+    implementation group: "org.glassfish.jersey.media", name: "jersey-media-multipart"
+    implementation group: "com.fasterxml.jackson.datatype", name: "jackson-datatype-jsr310"
+    implementation group: "io.swagger.core.v3", name: "swagger-annotations"
+    swaggerCodegen group: "io.swagger.codegen.v3", name: "swagger-codegen-cli"
```
Comment (fboulnois) [inline datarepo-client/build.gradle]: I am not sure exactly how to keep the original dependencies while using the new syntax, but we will otherwise have a conflict with some of the `unauthenticatedApi` generated code.

**Review by fboulnois — COMMENTED** [review]
Added my comments below:
